flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskExecutorStateChangelogStoragesManager.java

```diff
@@ -0,0 +1,143 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorage;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorageLoader;
+import org.apache.flink.util.ShutdownHookUtil;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class holds the all {@link StateChangelogStorage} objects for a task executor (manager). No
+ * thread-safe.
+ */
+public class TaskExecutorStateChangelogStoragesManager {
+
+    /** Logger for this class. */
+    private static final Logger LOG =
+            LoggerFactory.getLogger(TaskExecutorStateChangelogStoragesManager.class);
+
+    /**
+     * This map holds all state changelog storages for tasks running on the task manager / executor
+     * that own the instance of this. Maps from job id to all the subtask's state changelog
+     * storages.
+     */
+    private final Map<JobID, StateChangelogStorage<?>> changelogStoragesByJobId;
+
+    private boolean closed;
+
+    /** shutdown hook for this manager. */
+    private final Thread shutdownHook;
+
+    public TaskExecutorStateChangelogStoragesManager() {
+        this.changelogStoragesByJobId = new HashMap<>();
+        this.closed = false;
+
+        // register a shutdown hook
+        this.shutdownHook =
+                ShutdownHookUtil.addShutdownHook(this::shutdown, getClass().getSimpleName(), LOG);
+    }
+
+    public StateChangelogStorage<?> stateChangelogStorageForJob(
+            @Nonnull JobID jobId, Configuration configuration) {
+        if (closed) {
+            throw new IllegalStateException(
+                    "TaskExecutorStateChangelogStoragesManager is already closed and cannot "
+                            + "register a new StateChangelogStorage.");
+        }
+
+        StateChangelogStorage<?> stateChangelogStorage = changelogStoragesByJobId.get(jobId);
+
+        if (stateChangelogStorage == null) {
+            stateChangelogStorage = StateChangelogStorageLoader.load(configuration);
```

Comment:
(please see the comment above about returning `null` first)

If we stick with nulls then we'll be trying to load the storage each time, right? (even if it's not needed)

Reply:
Yeah. I think it's better to change the value type for map ```changelogStoragesByJobId``` as ```Reference<StateChangelogStorage<?>>```, then we could record the previous loaded ```null``` value. WDYT?

Reply:
Why `AtomicReference`, are there multiple threads accessing it? Wouldn't just `Optional` would be enough?

Reply:
Ah, it's my habit using ```AtomicReference```. ```Optional``` here is enough. I have changed this.
Thanks & sorry for this.

Reply:
Thanks!

I'm afraid it's a bit inconsistent now: we return `stateChangelogStorage.orElse(null)` from here and then pass it to `TaskStateManagerImpl` where it is marked `@Nonnull`.

Could you mark it `@Nullable`, and more importantly `TaskStateManager.getStateChangelogStorage`? (or use Optional)

Reply:
Right. I'll mark ```@ Nullable``` all the way to ```TaskStateManager.getStateChangelogStorage```.
Thanks for pointing this out 😄

Reply:
Thanks :slightly_smiling_face: